### PR TITLE
feat: ✨sd-carousel

### PR DESCRIPTION
**Pull request by Vahid1919**
## Description:
This PR includes the work for sd-carousel as well as sd-carousel item.

Related to https://github.com/solid-design-system/solid/issues/239

## Definition of Reviewable:
*PR notes: Irrelevant elements should be removed.*
- [x] Documentation is created/updated
- [x] Migration Guide is created/updated
- [x] E2E tests (features, a11y, bug fixes) are created/updated
- [x] Stories (features, a11y) are created/updated
- [x] relevant tickets are linked

### CODE REVIEW

File: packages/components/src/components/carousel-item/carousel-item.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
+import { css, html } from 'lit';
+import { customElement } from 'lit/decorators.js';
+import componentStyles from '../../styles/component.styles';
+import SolidElement from '../../internal/solid-element.js';
+
+/**
+ * @summary A carousel item represent a slide within a [carousel](/components/carousel).
+ *
```
Comment (MarcMatthiae) [inline packages/components/src/components/carousel-item/carousel-item.ts:8]:
```suggestion
 * @documentation https://solid.union-investment.com/[storybook-link]/carousel
 *
```

File: packages/components/src/internal/solid-element.ts [hunk 2]
```diff
@@ -34,6 +34,40 @@ export default class SolidElement extends LitElement {
...
+    }
+
+    // Need to make sure we're not working with null or empty strings before doing version comparisons.
+    if (newVersion && existingVersion && newVersion === existingVersion) {
+      // If versions match, we don't need to warn anyone. Carry on.
+      return;
+    }
+
+    console.warn(
+      `Attempted to register <${name}>${newVersion}, but <${name}>${existingVersion} has already been registered.`
```
Comment (mariohamann) [inline packages/components/src/internal/solid-element.ts:68]: Why is this here?

File: packages/components/src/styles/interactive/interactive.css [hunk 3]
```diff
@@ -16,6 +16,18 @@
...
+  &--inverted {
+    @apply text-white hover:text-primary-200;
+  }
+
+  &--inverted:disabled {
+    @apply text-neutral-600;
+  }
+
+  &--inverted:focus-visible {
+    @apply outline outline-2 outline-offset-2 outline-white;
```
Comment (mariohamann) [inline packages/components/src/styles/interactive/interactive.css:31]: Let's refactor this together.

File: packages/components/src/internal/solid-element.ts (hunk 2 above)
Comment (Vahid1919) [reply packages/components/src/internal/solid-element.ts:68]: This is was one of the dependencies I added for the first version of carousel. But I see it is no longer needed after we switched the version.   I can go ahead and remove it now.

File: packages/components/src/components/carousel-item/carousel-item.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+  render() {
+    return html` <slot></slot> `;
+  }
+
+  static styles = [
+    SolidElement.styles,
+    css`
+      ${componentStyles}
+      :host {
```
[the remark this answers is not in the document]
Comment (mariohamann) [reply packages/components/src/components/carousel-item/carousel-item.ts:36]: Yeah, should be possible. We could either add them globally to the solid-element.ts (as we do with tailwind.css) or import them component per component.

[the remark this answers is not in the document]
Comment (Vahid1919) [reply packages/components/src/components/carousel-item/carousel-item.ts:36]: @mariohamann @yoezlem I might need a little more clarification here. Do we want to use tailwind syntax here instead of the vanilla CSS I already have? 

I am not exactly sure how to add these specific styles since they're selecting for the host and not the slot itself(?)

[the remark this answers is not in the document]
Comment (mariohamann) [reply packages/components/src/components/carousel-item/carousel-item.ts:36]: Ah. Yeah so we don't have any styles that are relevant from our CSS styles here. So doing like you did is totally valid.

**Comment by Vahid1919** [comment]
@coraliefeil I've completed the sd-interactive changes and now awaiting approval from the Chromatic side :)

**Comment by Vahid1919** [comment]
> SRY cursor: pointer needed for pause / play button when hovering

added now

File: packages/components/src/components/carousel/carousel.test.ts [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+          await oneEvent(el.scrollContainer, 'scrollend');
+          await el.updateComplete;
+          // Act
+          await clickOnElement(nextButton);
+          await el.updateComplete;
+          // Assert
+          expect(nextButton).to.have.attribute('aria-disabled', 'true');
+          expect(el.next).not.to.have.been.called;
+        });
+        // TODO: This test times out.  Seems related to be `scrollend`.
```
Comment (karlbaumhauer) [inline packages/components/src/components/carousel/carousel.test.ts:278]: pls remove code here and add to fix-ticket if its blocking from merge as ToDos might get lost.

File: packages/components/src/components/carousel/carousel.test.ts [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+        await oneEvent(el.scrollContainer, 'scroll');
+        await el.updateComplete;
+
+        // Assert
+        expect(el.scrollContainer).to.have.attribute('aria-busy', 'true');
+
+        await oneEvent(el.scrollContainer, 'scrollend');
+        await el.updateComplete;
+
+        // TODO: This last assertion fails on Chromium. Seems related to `scrollend`.
```
Comment (karlbaumhauer) [inline packages/components/src/components/carousel/carousel.test.ts:545]: as in l278

File: packages/components/src/docs/Migration/ui-slider.mdx [hunk 7]
```diff
@@ -0,0 +1,153 @@
+# Migration Guide: From `ui-slider` to `sd-carousel`
+
+The new `sd-carousel` is designed to replace the `ui-slider`. Instead of mainly providing content via attributes, the `sd-carousel` component uses slots to allow for more flexibility and customization.
+
+## 💾 Slots
+
+### ✨ New Slots
+
+#### [default]
```
Comment (karlbaumhauer) [inline packages/components/src/docs/Migration/ui-slider.mdx:9]: istn't this the same as in the ui-slider? => would not be a new slot...

File: packages/components/src/styles/interactive/interactive.css [hunk 8]
```diff
@@ -4,18 +4,21 @@
     outline: revert;
```
Comment (karlbaumhauer) [inline packages/components/src/styles/interactive/interactive.css:4]: this should be in another PR as we use squash-merge in our PRs. This would result in one single commit with changes in 2 different components...

As other teams rely on our changelog, we need to be precise here.

File: packages/components/src/styles/interactive/interactive.declaration.ts [hunk 9]
```diff
@@ -10,6 +10,10 @@ export default {
       name: 'sd-interactive--disabled',
```
Comment (karlbaumhauer) [inline packages/components/src/styles/interactive/interactive.declaration.ts:10]: this should be in another PR as we use squash-merge in our PRs. This would result in one single commit with changes in 2 different components...

As other teams rely on our changelog, we need to be precise here.

File: packages/components/src/styles/interactive/interactive.stories.ts [hunk 10]
```diff
@@ -55,6 +55,27 @@ export const Disabled = {
   }
```
Comment (karlbaumhauer) [inline packages/components/src/styles/interactive/interactive.stories.ts:55]: this should be in another PR as we use squash-merge in our PRs. This would result in one single commit with changes in 2 different components...

As other teams rely on our changelog, we need to be precise here.

File: packages/components/src/styles/interactive/interactive.declaration.ts (hunk 9 above)
Comment (karlbaumhauer) [reply packages/components/src/styles/interactive/interactive.declaration.ts:10]: @mariohamann @yoezlem any thoughts on this topic?

**Review by karlbaumhauer — CHANGES_REQUESTED** [review]

File: packages/components/src/docs/Migration/ui-slider.mdx (hunk 7 above)
Comment (Vahid1919) [reply packages/components/src/docs/Migration/ui-slider.mdx:9]: Ah, you're right. Removed.

File: packages/components/src/styles/interactive/interactive.declaration.ts (hunk 9 above)
Comment (Vahid1919) [reply packages/components/src/styles/interactive/interactive.declaration.ts:10]: We decided to keep the declaration here since it was missing from #385 and it is only docs.

File: packages/components/src/components/carousel/carousel.stories.ts [hunk 11]
```diff
@@ -0,0 +1,212 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import '../../solid-components';
+import { html } from 'lit';
+import { storybookDefaults, storybookHelpers, storybookTemplate } from '../../../scripts/storybook/helper';
+import { userEvent } from '@storybook/testing-library';
+import { waitUntil } from '@open-wc/testing-helpers';
+
+const { argTypes, parameters } = storybookDefaults('sd-carousel');
+const { overrideArgs } = storybookHelpers('sd-carousel-item');
```
[the remark this answers is not in the document]
Comment (mariohamann) [reply packages/components/src/components/carousel/carousel.stories.ts:9]: @Vahid1919 Still open.

File: packages/components/tsconfig.json [hunk 12]
```diff
@@ -1,32 +1,45 @@
...
-    "forceConsistentCasingInFileNames": true,
-    "useUnknownInCatchVariables": true,
-    "baseUrl": ".",
-    "types": ["mocha", "user-agent-data-types", "vite/client"],
-    "resolveJsonModule": true
-  },
-  "include": ["src/**/*", "vite.config.*ts"],
-  "exclude": ["src/_components/**", "custom-elements-manifest.config.ts"]
+    "compilerOptions": {
+        /* Visit https://aka.ms/tsconfig.json to read more about this file */
```
[the remark this answers is not in the document]
Comment (Vahid1919) [reply packages/components/tsconfig.json]: I took a look at this. It seems like my auto formatter just ran through the file. I reverted the changes to make this PR more clear and the changes to the file more readable. 

--> The comment was missing too so I added it back.

**Comment by Vahid1919** [comment]
@coraliefeil @van-nguyen-ht Assigning this to you in regards to the UI Tests :)

**Comment by Vahid1919** [comment]
@solid-design-system/development Requesting re-review after resolving previous suggestions :)

**Comment by yoezlem** [comment]
Out of curiosity, what about all the smart devices? Is a swiper planned, or do users still need to click on the two arrows here?

**Comment by Vahid1919** [comment]
@yoezlem Scrolling is currently supported (eg: swiping with 2 fingers on the trackpad on Mac)
But I am not entirely sure if that works for smart devices. Mouse dragging specifically is not supported which might be the one for smart devices.

File: packages/components/src/components/icon/library.system.ts [hunk 13]
```diff
@@ -22,7 +22,15 @@ export const icons = {
...
   </svg>
-`
+`,
+  start: `
+  <svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="m19.54 11.16-14-9c-.153-.099-.339-.158-.54-.158-.552 0-.999.447-1 .998v18c0 .552.448 1 1 1 .201-.001.387-.06.544-.162l-.004.002 14-9c.271-.182.447-.487.447-.833s-.176-.652-.444-.831l-.004-.002zm-13.54 8.007v-14.333l11.15 7.167z"/></svg>
+  `,
+  pause: `
+    <svg viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
+    <path d="M9 22a.99997.99997 0 0 0 1-1V3a1.00003 1.00003 0 0 0-1-1H5a1.00003 1.00003 0 0 0-1 1v18a.99997.99997 0 0 0 1 1h4ZM6 4h2v16H6V4ZM19 22c.2652 0 .5196-.1054.7071-.2929A1.0001 1.0001 0 0 0 20 21V3a.99997.99997 0 0 0-1-1h-4a.99997.99997 0 0 0-1 1v18c0 .2652.1054.5196.2929.7071S14.7348 22 15 22h4ZM16 4h2v16h-2V4Z"/>
+      </svg>
```
Comment (mariohamann) [inline packages/components/src/components/icon/library.system.ts:32]: These are the correct new icon names, right?

File: packages/components/src/components/carousel/carousel.stories.ts [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+  args: overrideArgs([
+    {
+      type: 'slot',
+      name: 'default',
+      value: `
+        <sd-carousel-item> <slot-comp></slot-comp> </sd-carousel-item>
+        <sd-carousel-item> <slot-comp></slot-comp> </sd-carousel-item>
+        <sd-carousel-item> <slot-comp></slot-comp> </sd-carousel-item>
+        <sd-carousel-item> <slot-comp></slot-comp> </sd-carousel-item>
+        <sd-carousel-item> <slot-comp></slot-comp> </sd-carousel-item>`
```
Comment (mariohamann) [inline packages/components/src/components/carousel/carousel.stories.ts:24]: As you're currently working on the slot task – could you change this to BEM-styled divs, too?

**Review by mariohamann — CHANGES_REQUESTED** [review]

**Review by mariohamann — APPROVED** [review]

File: packages/components/src/components/carousel/carousel.stories.ts (hunk 14 above)
Comment (mariohamann) [reply packages/components/src/components/carousel/carousel.stories.ts:24]: When you're done: approved.

**Review by yoezlem — APPROVED** [review]

File: packages/components/src/components/carousel/carousel.stories.ts [hunk 15]
```diff
@@ -137,6 +137,28 @@ export const SlidesPerPage = {
   }
 };
 
+/**
+ * Use `slides-per-move` to set how many slides the carousel advances when scrolling. Useful when specifying a `slides-per-page`
```
Comment (karlbaumhauer) [inline packages/components/src/components/carousel/carousel.stories.ts:141]: what happens when `slides-per-page` is set to 1 and `slides-per-move` is set to 2? maybe you can explain the combination a bit more detailed in terms of what to expect => makes it easier to test.

File: packages/components/src/components/carousel/carousel.ts [hunk 16]
```diff
@@ -65,6 +65,12 @@ export default class SdCarousel extends SolidElement {
   /** Specifies how many slides should be shown at a given time.  */
   @property({ type: Number, attribute: 'slides-per-page' }) slidesPerPage = 1;
 
+  /**
+   * Specifies the number of slides the carousel will advance when scrolling, useful when specifying a `slides-per-page`
```
Comment (karlbaumhauer) [inline packages/components/src/components/carousel/carousel.ts:69]: if changed in the stories, pls also change here.

**Review by karlbaumhauer — CHANGES_REQUESTED** [review]

File: packages/components/src/components/icon/library.system.ts (hunk 13 above)
Comment (Vahid1919) [reply packages/components/src/components/icon/library.system.ts:32]: Yes!

File: packages/components/src/components/carousel/carousel.stories.ts (hunk 15 above)
Comment (Vahid1919) [reply packages/components/src/components/carousel/carousel.stories.ts]: @karlbaumhauer I added some explanation there, but I thought it was better to show how it works so I expanded on it's story to show that specific example.

File: packages/components/src/components/carousel/carousel.ts (hunk 16 above)
Comment (Vahid1919) [reply packages/components/src/components/carousel/carousel.ts]: Updated here as well.

**Comment by Vahid1919** [comment]
@karlbaumhauer @mariohamann 
I have updated the documentation for SlidesPerMove detailing how it should be used and how it should not. Please Review :)

**Review by karlbaumhauer — APPROVED** [review]

**Comment by karlbaumhauer** [comment]
@Vahid1919 pls change the version in the docs as we already are at v1.17.0 and the sd-carousel will be stable at v1.18.0 afterwards.

**Review by mariohamann — APPROVED** [review]
